Add filterprocessor for dropping metrics and traces

anniefu opened the pull request:
**Description:** Add "filterprocessor" to allow dropping metrics and traces completely from the Collector's pipeline. Currently only filters by metric name or trace name. See updated processor README.md for usage options.

- Only one of include/exclude can be specified, but filter processors can be chained
- Filter can be uniquely configured per trace/metric field to filter on (currently just name)
- Optional LRU caching for regexp mode

**Link to tracking Issue:** #297 

**Testing:** unit testing for each component & local e2e testing for Knative w/ opencensus agent -> OpenTelemetry Collector Service -> Stackdriver

**Documentation:** "Filter Processor" section of processor's README.md

owais commented:
I didn't review in detail but any reason this is not being added to the contrib repo instead (https://github.com/open-telemetry/opentelemetry-collector-contrib) ? Also, not sure about splitting the processor into two packages and putting one in internal.

anniefu commented:
Added here instead of contrib repo because it is a feature request for core (#297). I am okay with adding it to contrib repo instead if we decide that's the correct place for it.

The internal processor package does not actually contain a processor, just helper libraries for processors. The attributesprocessor was already using that internal/processor for that purpose, so I added my helper library there. Happy to place it under processor/filterprocessor/internal too though.

anniefu commented:
Hi Tigran, thanks for the suggestions! I will bring the design discussion #297 with a longer comment to explain the design in this PR and certain decisions.

I can break this down into smaller PRs after that discussion.
